### ISSUE TITLE
Errors in the Adam implementation

### ISSUE THREAD

**Issue by ChristopherBrix** [mila-udem/blocks#1175]
I think I discovered some errors in the current implementation of adam.

Together, they cause the learning to fail (tested on a relatively small training set). First, the average cross entropy decreases, but after some epochs (20 in my case) it increases again and soon (after 29 epochs) only `nan` is reported. 

- The initial `learning_rate` ([0.002](https://github.com/mila-udem/blocks/blob/master/blocks/algorithms/__init__.py#L861)) doesn't match the value provided in the [current version 8 of the paper](https://arxiv.org/pdf/1412.6980v8.pdf) (0.001)
I'm aware of #767, but I don't know about the sources for the new proposed learning rate. In my opinion, the default value should match the values used in the paper.
- Adam moment parameterization is backwards (#1159). 
- [The name of `beta_1t`](https://github.com/mila-udem/blocks/blob/master/blocks/algorithms/__init__.py#L880) is misleading in my eyes. It is the decayed value of `beta_1`, but not to the power of `t`
- The learning rate isn't decayed as described in the paper (above Theorem 4.1)
- The influence of `epsilon` is increased, because it's [implicitly divided](https://github.com/mila-udem/blocks/blob/master/blocks/algorithms/__init__.py#L882) by `sqrt(1-b_2**t)`. This could be fixed by explicitly calculating `variance_t_hat`.

After I fixed all these issues in a local version of adam, I was able to train the network (slowly, but) successfully.

This is my current version of Adam:
```
def __init__(self, learning_rate=0.001,
             beta1=0.9, beta2=0.999, epsilon=1e-8,
             decay_factor=(1 - 1e-8)):
    self.learning_rate = shared_floatx(learning_rate, "learning_rate")
    self.beta1 = shared_floatx(beta1, "beta1")
    self.beta2 = shared_floatx(beta2, "beta2")
    self.epsilon = shared_floatx(epsilon, "epsilon")
    self.decay_factor = shared_floatx(decay_factor, "decay_factor")
    for param in [self.learning_rate, self.beta1, self.beta2, self.epsilon,
                  self.decay_factor]:
        add_role(param, ALGORITHM_HYPERPARAMETER)

def compute_step(self, parameter, previous_step):
    logging.warning("Using custom adam code")
    mean = _create_algorithm_buffer_for(parameter, 'mean')
    variance = _create_algorithm_buffer_for(parameter, 'variance')
    time = shared_floatx(0., 'time')
    add_role(time, ALGORITHM_BUFFER)

    t1 = time + 1
    beta_1_decayed = self.beta1 * self.decay_factor ** (t1 - 1)
    learning_rate_decayed = self.learning_rate / tensor.sqrt(t1)
    mean_t = beta_1_decayed * mean + (1. - beta_1_decayed) * previous_step
    variance_t = self.beta2 * variance + \
                  (1. - self.beta2) * tensor.sqr(previous_step)

    mean_t_hat = mean_t / (1. - beta_1_decayed ** t1)
    variance_t_hat = variance_t / (1. - self.beta2 ** t1)

    step = (learning_rate_decayed * mean_t_hat /
            (tensor.sqrt(variance_t_hat) + self.epsilon))

    updates = [(mean, mean_t),
               (variance, variance_t),
               (time, t1)]

    return step, updates
```

If you agree that my modifications are correct, I'll happily create a pull request so they can be incorporated.

**Comment by SwordYork** [comment]
Hi, ChristopherBrix,

Since I have modified the code, I would like to clarify something.

- I think simply setting learning rate to 0.001 could avoid nan in your case.
- Decaying learning rate as the paper may slowdown the training process.
- mean_t_hat and variance_t_hat should't used in the implementation, you could refer Section 2 of the paper or [the implemention in Keras](https://github.com/fchollet/keras/blob/master/keras/optimizers.py#L427). 

> Note that the efficiency of algorithm 1 can, at the expense of clarity, be improved upon by changing
> the order of computation, e.g. by replacing the last three lines in the loop with the following lines:
> $ \alpha_t = \alpha\sqrt{1-\beta_2^t}/(1-\beta_1^t) $ and $\theta_t \gets \theta_{t-1} - \alpha_t m_t / (\sqrt{v_t} + \epsilon)$.
> 
It is equivalent, but more efficient.

**Comment by dmitriy-serdyuk** [comment]
I remember that the Adam implementation first was basically copied from a gist by Kingma. Then it was changed and refactored. The paper has changed 7 times on arxiv since then and I have no idea if a new Adam works better than the original one.

Personally, I haven't seen big problems with Adam implementation apart from crazy models where any optimizer works badly.

@ChristopherBrix when you say slowly, do you mean the convergence speed or the computation speed?

**Comment by ChristopherBrix** [comment]
Thank you for you quick replies!

@SwordYork:
 - Do you think this only affects my case, or should the default value be changed generally?
 - Yes, after removing the decay the network learned much faster. I found a [comment by D. P. Kingma](https://github.com/mila-udem/blocks/pull/588#commitcomment-11292038) stating that "The exponential decay of beta1 was added for theoretic reasons" - if this does extend to the decay of the learning rate, I agree that it shouldn't be used
 - In the paper, they switch to `epsilon_hat` when they do this optimization. It isn't explicitly defined, but I guess it should be `epsilon_hat = epsilon * sqrt(1-b_2**t)` because than the implicit division would cancel out. I agree that the influence is rather small (after 40 epochs, epsilon increased by a factor of 5), but this should be either fixed or documented somewhere

@dmitriy-serdyuk: It converged really slowly (I get a cross entropy (on the training set) of 47 after the second epoch with the current implementation, my implementation took 47 epochs to decline to a cross entropy of 70). But the computation speed was actually a bit better (2.35 epochs per hour vs. only 2.05 epochs per hour) but that's probably due to some fluctuation on the cluster I use.

**Comment by SwordYork** [comment]
Hi, @ChristopherBrix 

- I'm not sure, it depends on the model.
- I think `epsilon` is mainly for numerical stability, it should not relate to the convergence and should not be changed during optimization. Thus, either `epsilon`  or `epsilon_hat` should be fixed to a small positive number before training.

**Comment by rizar** [comment]
My 5 cents:

- I don't think it makes sense to try and figure out the best learning rate. For different models different learning rates work differently well.

- The learning rate decay indeed seems to be used in the paper mostly for theoretical reasons. I think the same is probably true about decay of beta_1 (and it seems like Durk confirmed it). The latter at least doesn't hurt when lambda is very close 1, whereas the learning rate decay as described in the paper could be very harmful.

- You are probably right about epsilon. At the same time, the difference between \hat{epsilon} and epsilon will decrease over time. Also, other implementations seems to handle epsilon in exactly the same way, e.g. [tensorflow](https://github.com/tensorflow/tensorflow/blob/7c16e2e23c4d1234903db97ada36f870b724c71d/tensorflow/core/kernels/training_ops_gpu.cu.cc)

- And yes, we should change the way the moments are parametrized. I would be willing to accept the respective PR.

**Comment by ChristopherBrix** [comment]
@rizar: Oh, you're right, the influence of the implicit division *decreases* over time. So the difference is the worst in the first epoch, when epsilon is increased by a factor of 31. I don't know whether that's a significant increase or not.

**Comment by dwf** [comment]
Note that 1-1e-8 in float32 is 1.0, so at some point (and maybe now), the
default decay for both our implementation and Durk's was 1.0.

>>> numpy.float32(1) - numpy.float32(1e-8) == numpy.float32(1.0-1e-8) == 1
True

So I don't think it's worth worrying about.

**Comment by ChristopherBrix** [comment]
Oh, I didn't think about that. Maybe we should change `1.0 - 1e-8` to `1`? Then no-one else would get confused.

**Comment by rizar** [comment]
Yeah, we should do it. Please feel free to create a PR.

**Comment by ChristopherBrix** [comment]
`numpy.float64(1) - numpy.float64(1e-8) == numpy.float64(1.0-1e-8) == 1` is `False`, so we can't simplify the decay to `1`. If a user configures theano to use `float64`, the decay sets in. This is currently tested: https://travis-ci.org/mila-udem/blocks/jobs/194762433

If this is the intended behavior, I'd recommend documenting this somewhere. Otherwise the test will have to be modified.
